src/System/IO/ByteBuffer.hs

```diff
@@ -92,10 +103,52 @@ data BBRef = BBRef {
       -- the 'ByteBuffer'
     }
 
-type ByteBuffer = IORef BBRef
+-- | Exception that is thrown when an invalid 'ByteBuffer' is being used that is no longer valid.
+--
+-- A 'ByteBuffer' is considered to be invalid if
+--
+-- - it has explicitly been freed
+-- - an Exception has occured during an operation that modified it
+data ByteBufferException = ByteBufferException
+    { _bbeLocation :: !String
+      -- ^ function name that caused the exception
+    , _bbeException :: !String
+      -- ^ printed representation of the exception
+    }
+    deriving (Typeable, Eq)
+instance Show ByteBufferException where
+    show (ByteBufferException loc e) = concat
+        ["ByteBufferException: ByteBuffer was invalidated because of Exception thrown in "
+        , loc , ": ", e]
+instance Exception ByteBufferException
+
+type ByteBuffer = IORef (Either ByteBufferException BBRef)
+
+-- | On any Exception, this will invalidate the ByteBuffer and re-throw the Exception.
+--
+-- Invalidating the 'ByteBuffer' includes freeing the underlying pointer.
+bbHandler ::
```

Comment:
This function is always use paired with `handle`. I'd just include the call to `handle` here.